flang/lib/Lower/ConvertCall.cpp

```diff
@@ -1232,8 +1232,28 @@ genUserCall(Fortran::lower::PreparedActualArguments &loweredActuals,
 
   if (!fir::getBase(result))
     return std::nullopt; // subroutine call.
-  // TODO: "move" non pointer results into hlfir.expr.
-  return extendedValueToHlfirEntity(loc, builder, result, ".tmp.func_result");
+
+  hlfir::Entity resultEntity =
+      extendedValueToHlfirEntity(loc, builder, result, ".tmp.func_result");
+
+  if (!fir::isPointerType(fir::getBase(result).getType())) {
+    // Just load trivial scalars right away.
+    if (resultEntity.isScalar() &&
+        fir::isa_trivial(resultEntity.getFortranElementType()))
+      resultEntity = loadTrivialScalar(loc, builder, resultEntity);
+
+    if (resultEntity.isVariable()) {
+      // Function result must not be freed, since it is allocated on the stack.
```

Comment:
What about `hlfir.destroy`? I think in other cases we do create a destroy operation even when the result type does not need deallocation. But I am open to not doing so.

Reply:
I think it should be possible to use `hlfir.destroy`, though, currently we destroy the function result as an in-memory entity (e.g. see https://github.com/llvm/llvm-project/pull/67768).  It may even be a potential correctness issue currently, e.g. if the `as_expr` is moved along with its use past the destroy call.  @jeanPerier, do you think we need to make the buffer live-range explicit here with `hlfir.as_exp`->`hlfir.destroy` def-use?

Reply:
Tom, it makes sense to use `hlfir.destroy` for the clean-up of the non-elemental result.  I will do it, when we get rid of the FIR path.  I hope using as_expr now should not result in incorrect MLIR optimizations.

Reply:
> It may even be a potential correctness issue currently, e.g. if the as_expr is moved along with its use past the destroy call. @jeanPerier, do you think we need to make the buffer live-range explicit here with hlfir.as_exp->hlfir.destroy def-use?

Yes, it would be cleaner. As you mentioned, the clean-up is currently done on the memory that was available before it is moved, this probably does not cause any issue currently, but could if the life of hlfir.expr is prolongated for whatever reasons. I agree the change will be easier without having to deal with both FIR and HLFIR.